sympy/sets/fancysets.py

```diff
@@ -910,14 +909,27 @@ def _boundary(self):
 
     def as_relational(self, x):
         """Rewrite a Range in terms of equalities and logic operators. """
-        from sympy.functions.elementary.integers import floor
         if self.size == 1:
             return Eq(x, self[0])
+        elif self.size == 0:
+            return S.false
         else:
-            return And(
-                Eq(x, floor(x)),
-                x >= self.inf if self.inf in self else x > self.inf,
-                x <= self.sup if self.sup in self else x < self.sup)
+            from sympy.core.mod import Mod
+            cond = None
+            if self.start.is_infinite:
+                if self.stop.is_infinite:
+                    cond = S.true
+                else:
+                    a = self.reversed.start
+            elif self.start == self.stop:
+                cond = S.false  # null range
+            else:
+                a = self.start
+            step = abs(self.step)
+            cond = Eq(Mod(x, step), a % step) if cond is None else cond
+            return And(cond,
```

Comment:
I'm not sure if we need to require values to be integers for Range...if we do, then the condition `Eq(floor(a), a)` must be added. But I'm not convinced that it needs to be restricted to Integers. Our only requirement is that we be able to compute the size. `Range(sqrt(2), 5, sqrt(3))` is well defined as `[sqrt(2), sqrt(2) + sqrt(3)]`.

Reply:
Let's just add the condition that a must be an integer: `Eq(Mod(a, 1), 0)`. When dealing with non-symbolic ranges this will default to True.

Reply:
>compute the size. `Range(sqrt(2), 5, sqrt(3))` is well defined as `[sqrt(2), sqrt(2) + sqrt(3)]`.

Yes this is logically accurate, But the code for Range does not accept `sqrt` as argument , instead it raises an error.
I was also thinking can `Range(1.2 , 5 )`  can be defined as `[1.2 , 2.2 , 3.2 , 4.2]` if all of this is correct than that means there is still so much to modify. And different things to work on Range.

Reply:
> Let's just add the condition that a must be an integer: `Eq(Mod(a, 1), 0)`. When dealing with non-symbolic ranges this will default to True.

In the code from line 600 to 609 Range class already deals with the non integer inputs , If I try to add the condition (`Eq(Mod(a,1),0)`) as a condition for `a` to be an integer ,
 than there will be many major changes and we have to changes a lot of things other than this.

```python
 elif not w.is_Integer:
                    raise ValueError
                else:
                    ok.append(w)
        except ValueError:
            raise ValueError(filldedent('''
    Finite arguments to Range must be integers; `imageset` can define
    other cases, e.g. use `imageset(i, i/10, Range(3))` to give
    [0, 1/10, 1/5].'''))
```

Reply:
The problem is that we can't give a result based on the Range object that does not give the same answer as the non-symbolic form of the range. So although `Range(x, x + 2*y, y)` is ok, inf and sup are not known even though the size is known and the relational expression should raise an error since there is little point in generating a large relational expression to cover all the possibilities. But for now, let's stop here with this step forward.